text/0321-key-functions.md

```diff
@@ -0,0 +1,74 @@
+- Start Date: 2018-03-24
+- RFC PR: https://github.com/emberjs/rfcs/pull/321
+- Ember Issue: (leave this empty)
+
+# Key Functions for `each`/`each-in`
+
+## Summary
+
+This RFC proposes that we support passing functions to the `key` argument of the `each` and `each-in` helpers. These functions would be called for each item and would be expected to return they key value used in diffing.
+
+For example,
+
+```js
+import Component from '@ember/component';
+
+export default Component.extend({
+  itemKey(item) {
+    return `${item.factoryNumber}-${item.partNumber}`;
+  }
+});
+```
+
+```hbs
+{{#each @items key=itemKey as |item|}}
+  {{item-details item=item}}
+{{/each}}
+```
+
+## Motivation
+
+The `key` argument to `each` and `each-in` determines the key used in the Glimmer diffing algorithm. Here is a description of the currently accepted values?
+
+`key` argument          | Computed key
+------------------------|---------------------------------------------
+`@identity`             | `guidFor(item)`
+`@index`                | index of the item in the array / object keys
+`@key` (`each-in` only) | the key of the key-value pair
+other strings           | `get(item, path)`
+otherwise               | error
+
+If a key is not specified then `@identity` strategy is used.
+
+Unfortunately, unless one of the default strategies works for your use case, you're stucking using a path on your object as the key. This can be problematic when you need to assemble a composite key from multiple values as in the example above. It is especially problematic when dealing with objects whose fields are out of your control (e.g. a moment date, an immutable.js record, a frozen object, etc.). Another less common reason you might want to use a key function is if your keys are expensive to compute and you want to cache them in a `WeakMap`.
+
+## Detailed design
+
+If `key` is a function, then Glimmer's diffing algorithm will call this function to determine the item's key.
+
+For `each`, the function would be called with arguments `keyFn(item, index)`.
+For `each-in`, the function would be called with arguments `keyFn(key, value)`.
+
+The key function should not be invoked with a `this`.
```

Comment:
what does that mean?

Reply:
That glimmer should call the key function as `keyFn(...args)` as not to bind `this` to any particular value.

Reply:
but with https://github.com/emberjs/rfcs/pull/308 would we then need to use `key=this.someUnboundFunction`? would seem strange if that was actually not bound to `this` in that case 🤔

Reply:
I don't think it's strange. For example,

```hbs
{{some-component onClick=this.someUnboundFunction}}
```

is also not bound. This is generally true (albeit a bit annoying!) in Handlebars today.

Reply:
yeah, true, I guess passing around `this.someUnboundFunction` in JS also doesn't have to mean that it's bound...

Reply:
> The key function should not be invoked with a `this`.

I think due to #308 we should rephrase this as something like "The key function will not be bound to `this` automatically", since in the future we will have to call it with `this` if #308 gets merged.

Reply:
:+1: That phrasing is better.